[BEAM-6231] Make Dataflow runner harness work with FixedWindow

boyuanzz opened the pull request:
There is a window error "IntervalWindow cannot be cast to GlobalWindow" introduced by https://github.com/apache/beam/pull/7015. This PR fixes this error and removes all DataflowPortabilityExecutableStageUnsupported tags. 
This test report https://scans.gradle.com/s/645npxy7rmibi/tests is running with
```
includeCategories 'org.apache.beam.testing.DataflowPortabilityExecutableStageUnsupported'
excludeCategories 'org.apache.beam.testing.UsesSideInputs'
excludeCategories 'org.apache.beam.testing.UsesStatefulPardo'
```
R: @robertwb 
cc: @HuangLED

boyuanzz commented:
Run Python PreCommit

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -163,7 +167,7 @@ public Node apply(MutableNetwork<Node, Edge> input) {
 
-    // Use default WindowingStrategy as the fake one.
+    // Use default WindowingStrategy as the fake one. But it may be overrided later if it's a FixedWindow.
     // TODO: should get real WindowingStategy from pipeline proto.
```
I agree and getting pipeline proto from Dataflow service should be an ultimate solution. But it's kind of hard for me to change the proto from Dataflow service, and the proto changes will also take some time. This generated WindowingStrategy is a short term workaround for now. We are discussing the possibility of changing proto from Dataflow service.

robertwb replied:
So the v1beta3 protos expose neither the windowing strategy nor a reference that can be used to tie back to the original PCollection? Unfortunate...

In this case, let's add a (possibly memoized) method `getWindowingStrategyForCoder(Coder<?> windowCoder, componentsBuilder)` rather than creating and injecting a default and then later possibly overwriting it. It can have a case for Global and Interval windows, and throw an error otherwise.

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
+        Coder<?> javaCoder =
+            CloudObjects.coderFromCloudObject(CloudObject.fromSpec(instructionOutput.getCodec()));
+        // For now, Dataflow runner harness only deal with FixedWindow.
+        if (FullWindowedValueCoder.class.isInstance(javaCoder)) {
```
For ```instanceof``` vs ````class.isInstance```, ```instanceof``` checks the type during compiling time but ```class.isInstance``` checks the class of object dynamically (https://www.geeksforgeeks.org/instanceof-operator-vs-isinstance-method-in-java/).

boyuanzz replied:
For the error, it means there is something wrong when converting a FixWindows into proto. If it fails to convert, then the communication should be broken owing to incomplete proto.

robertwb replied:
What I meant was is there any case where javaCoder would not be an instance of FullWindowedValueCoder?

As for isinstance, FullWindowedValueCoder.class is not dynamic, so there's no need to use the less-common form here.  (Same below.)

boyuanzz replied:
From what I know so far, if there is a window, then FullWindowedValueCoder is the only coder. I feel like the coder I got from v1b3 should be a FullWindowedValueCoder.

robertwb replied:
In that case, let's error if not. (Or otherwise discover the exception and properly handle it.)

boyuanzz commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
         Iterables.filter(input.nodes(), InstructionOutputNode.class)) {
       InstructionOutput instructionOutput = node.getInstructionOutput();
 
+      // If this is the input PCollection or the output PCollection for an ExecutableStage, it's
```
From my understanding, the input and output PCollection go through fnapi data channel between runner harness and sdk harness. Taking sending value from runner harness as an example, it needs a coder(```element coder``` + ```window coder```) to encode value then send it to sdk harness. The ```element coder``` is got from component by ```coderId```, and ```window coder``` is got from WindowingStrategy by ```windowingStrategyId```.

robertwb replied:
Sorry, let me rephrase this. Is there a reason we should avoid this for non-input-output PCollections?

boyuanzz replied:
I'm not sure whether I understand the coder usage correctly, but feel like, only the PCollection which is across the sdk-runner boundary cares the window coder. Thus, it's ok just assign the GlobalWindow to these non-input-output PCollections.

robertwb replied:
I'd rather not selectively lie, unless there's a good reason to do so.

robertwb commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
+              componentsBuilder.putWindowingStrategies(
+                  fakeWindowingStrategyId, windowingStrategyProto.getWindowingStrategy());
+              componentsBuilder.putAllCoders(windowingStrategyProto.getComponents().getCodersMap());
+              componentsBuilder.putAllEnvironments(
```
Do we need this? Wouldn't this break for non-Java?

boyuanzz replied:
I didn't test these against non-java but I guess there is a huge possibility to break non-java. This is another concern, I may not be able to extract window coder and element coder from a non-java coder.

robertwb commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
+          if (IntervalWindowCoder.class.isInstance(windowCoder)) {
+            fakeWindowingStrategyId = "generatedFixedWindowingStrategy" + idGenerator.getId();
+            try {
+              // Since the coder is the only needed from a WindowingStrategy, the size field of one
```
Is this WindowingStrategy instantiated in the SDK? If not, I'd much rather provide an actual fake strategy that has nothing but the coder set (or perhaps a nonce for the URN) to make sure it doesn't actually get used.

boyuanzz replied:
Ideally, the WindowingStrategy should be instantiated in the SDK and passed with pipeline proto to runner harness. However, dataflow runner harness doesn't have pipeline proto thus dataflow generates this fake one in order to make it work with fnapi.

robertwb replied:
Yes, I agree. I'm suggesting we create a very fake one, manually setting the coder_id to the right thing (regardless of whether it's understood by Java), rather than a FixedWindows.

boyuanzz replied:
Could you please explain more about how we can manually set the coder_id to the right thing?

boyuanzz replied:
Got it. Thanks~

boyuanzz replied:
If we don't want to lie on the internal PCollection, then it's not possible to build a totally fake WindowingStrategy. For internal PCollections inside one ExecutanleStage, it needs to initialize WindowFn from WindowingStrategy proto.

HuangLED commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -191,6 +195,37 @@ public Node apply(MutableNetwork<Node, Edge> input) {
+      // If this is the input PCollection or the output PCollection for an ExecutableStage, it's
+      // necessary to check whether the window coder is not a GlobalWindow coder.
+      if (isExecutableStageInputPCollection(input, node)
+          || isExecutableStageOutputPCollection(input, node)) {
```
for education purpose, what are the pcollection types beyond Input and output?

boyuanzz replied:
As for an ExecutableStage, there are 3 types of PCollection from my knowledge. The ```input PCollection``` is the PCollection flowing from runner to sdk. The ```output PCollection``` is the PCollection flowing from sdk to runner. The other PCollections are the intermediate PCollection between 2 PTransforms.

boyuanzz commented:
Run Seed Job

boyuanzz commented:
commit https://github.com/apache/beam/pull/7356/commits/4f3eaf51db8959e3563fafa786c4305352a08896 addressed the comments above:

- The function will register GlobalWindowingStrategy and FixedWindowingStrategy into componentBuilder before creating all PCollections.
- Instead of only caring about input and output PCollections, it will put the right kind of WindowingStrategy based on window coder to all PCollection. This logic only applies to java coder, for python, it still uses GlobalWindow as default.

Test log: https://scans.gradle.com/s/kameqe3xqgg4u/tests

@robertwb Could you please review this latest changes?

robertwb commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -167,19 +167,24 @@ public Node apply(MutableNetwork<Node, Edge> input) {
 
-    // Use default WindowingStrategy as the fake one. But it may be overrided later if it's a FixedWindow.
+    // By default, use GlobalWindow for all languages.
+		// For java, if there is a IntervalWindowCoder, then use FixedWindow instead.
```
Indentation.

robertwb commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -167,19 +167,24 @@ public Node apply(MutableNetwork<Node, Edge> input) {
+		// For java, if there is a IntervalWindowCoder, then use FixedWindow instead.
     // TODO: should get real WindowingStategy from pipeline proto.
-    String fakeWindowingStrategyId = "fakeWindowingStrategy" + idGenerator.getId();
+    String fakeGlobalWindowingStrategyId = "fakeGlobalWindowingStrategy" + idGenerator.getId();
```
Is this one actually fake? Perhaps just call it globalWindowingStrategyId.

robertwb commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -167,19 +167,24 @@ public Node apply(MutableNetwork<Node, Edge> input) {
     // TODO: should get real WindowingStategy from pipeline proto.
-    String fakeWindowingStrategyId = "fakeWindowingStrategy" + idGenerator.getId();
+    String fakeGlobalWindowingStrategyId = "fakeGlobalWindowingStrategy" + idGenerator.getId();
+    String fakeFixedWindowingStrategyId = "fakeFixedWindowingStrategy" + idGenerator.getId();
```
You could call this intervalWindowEncodingWindowingStrategyId or similar to clearly identify the properties you need to leverage here.

robertwb commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -235,6 +210,17 @@ public Node apply(MutableNetwork<Node, Edge> input) {
           RunnerApi.Coder coderProto = CoderTranslation.toProto(elementCoder, sdkComponents);
           componentsBuilder.putCoders(coderId, coderProto);
+          // For now, Dataflow runner harness only deal with FixedWindow.
+          if (FullWindowedValueCoder.class.isInstance(javaCoder)) {
```
You didn't address my comment of why you're not doing `javaCoder isinstance FullWindowedValueCoder`. Generally classObject.isInstrance(Object) is only needed if classObject is not known at compile time.

robertwb commented on runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/graph/CreateExecutableStageNodeFunction.java:
```diff
@@ -195,38 +200,8 @@ public Node apply(MutableNetwork<Node, Edge> input) {
-      }
-
       String coderId = "generatedCoder" + idGenerator.getId();
+      String windowingStrategyId = fakeGlobalWindowingStrategyId;
```
I would prefer being defensive and leave this unset, then below setting it in each of the conditions (including adding a check for global windows and the "Coder probably wasn't a java coder" exception clause. This will make it clearer when we're right and when we're just sticking something in.

boyuanzz replied:
I don't think we should have an exception clause if this is not a java coder. In my latest changes, I set it to use GlobalWindow if it's not a java coder.

boyuanzz commented:
Please take another look at commit https://github.com/apache/beam/pull/7356/commits/7857ab0886275d6170f7e189048406baaf2f10a4, @robertwb . Thanks so much for your help!

asfgit commented:
SUCCESS 
 
--none--

boyuanzz commented:
Run Java PreCommit

boyuanzz commented:
Run Java PreCommit
